include/dlaf/eigensolver/band_to_tridiag/mc.h

```diff
@@ -101,55 +100,160 @@ class BandBlock {
 
-  void copyOffdiag(SizeType j, const ConstTileType& source) noexcept {
-    constexpr auto General = blas::Uplo::General;
-    constexpr auto Upper = blas::Uplo::Upper;
-    // The elements are copied in the following way:
-    // (a: copied with first lacpy (Upper), b: copied with second lacpy (General))
-    // (copied when j = n)
-    // 6x6 tile, band_size = 3  |  2x6 tile, band_size = 3
-    // * * * a a a              |
-    // * * * * a a              |  * * * a a b
-    // * * * * * a              |  * * * * a b
-    // * * * * * *              |
-    // * * * * * *              |
-    // * * * * * *              |
-    const auto index = source.size().cols() - band_size_;
-    const auto size = std::min(band_size_, source.size().rows());
-    auto dest = ptr(band_size_, j + index);
-    lapack::lacpy(Upper, size, size, source.ptr({0, index}), source.ld(), dest, ld());
-    if (band_size_ > size) {
-      const auto size2 = band_size_ - size;
-      lapack::lacpy(General, source.size().rows(), size2, source.ptr({0, index + size}), source.ld(),
-                    dest + ld() * size, ld());
+  template <Device D, class Sender>
+  auto copyOffDiag(const SizeType j, Sender source) noexcept {
+    using dlaf::internal::transform;
+
+    namespace ex = pika::execution::experimental;
+
+    constexpr auto B = dlaf::matrix::internal::CopyBackend_v<D, Device::CPU>;
+
+    if constexpr (D == Device::CPU) {
+      return transform(
+          dlaf::internal::Policy<B>(),
+          [=](const matrix::Tile<const T, D>& source) {
+            constexpr auto General = blas::Uplo::General;
+            constexpr auto Upper = blas::Uplo::Upper;
+            // The elements are copied in the following way:
+            // (a: copied with first lacpy (Upper), b: copied with second lacpy (General))
+            // (copied when j = n)
+            // 6x6 tile, band_size = 3  |  2x6 tile, band_size = 3
+            // * * * a a a              |
+            // * * * * a a              |  * * * a a b
+            // * * * * * a              |  * * * * a b
+            // * * * * * *              |
+            // * * * * * *              |
+            // * * * * * *              |
+            const auto index = source.size().cols() - band_size_;
+            const auto size = std::min(band_size_, source.size().rows());
+            auto dest = ptr(band_size_, j + index);
+            ::lapack::lacpy(Upper, size, size, source.ptr({0, index}), source.ld(), dest, ld());
+            if (band_size_ > size) {
+              const auto size2 = band_size_ - size;
+              ::lapack::lacpy(General, source.size().rows(), size2, source.ptr({0, index + size}),
+                              source.ld(), dest + ld() * size, ld());
+            }
+          },
+          std::move(source));
+    }
+#ifdef DLAF_WITH_CUDA
+    else if constexpr (D == Device::GPU) {
+      DLAF_ASSERT_HEAVY(isAccessibleFromGPU(), "BandBlock memory should be accessible from GPU");
```

Comment:
It's a good point. This assertion has been added to ensure that CPU memory allocated by BandBlock for the workspace matrix is accessible from the GPU, so we can work with lacpy GPU kernel to copy directly from GPU to CPU.

GPU can access CPU memory iff it is pinned memory.

Let's check how we allocate host memory in DLAF.

https://github.com/eth-cscs/DLA-Future/blob/a285bd501c852e0b275e77826bb95884a69f2d10/include/dlaf/memory/memory_chunk.h#L63-L65

https://github.com/eth-cscs/DLA-Future/blob/a285bd501c852e0b275e77826bb95884a69f2d10/src/memory/memory_chunk.cpp#L20-L24

Since BandBlock internal CPU memory is pinned, it is accessible from GPU.

As already said, the assertion is just for ensuring this fact, since it is not a given fact that CPU memory is accessible from GPU, and the place where this memory has been made "valid" is unrelated to this code.

For the assertion I used [cudaPointerGetAttributes](https://docs.nvidia.com/cuda/archive/11.0/cuda-runtime-api/group__CUDART__UNIFIED.html#group__CUDART__UNIFIED_1gd89830e17d399c064a2f3c3fa8bb4390). Whatever type of memory it returns that is not `cudaMemoryTypeUnregistered` should be ok for the purpose, since:
- `cudaMemoryTypeHost`: it is registered host memory, so it is "page-locked"/"pinned"
- `cudaMemoryTypeDevice`: it is already on gpu, so it is accessible trivially
- `cudaMemoryTypeManaged`: it is managed memory, i.e. unified memory accessible by all processors

About first point with registered memory, I'm not sure how to deal with that in case of multiple GPUs (see [cudaHostRegister](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1ge8d5c17670f16ac4fc8fcb4181cb490c)). Probably in this case won't be enough.

Do you have any suggestion?

Reply:
Today I learned! I only knew pinned host memory was required for asynchronous transfers, but not that it could actually be accessed directly on the device. Are there any special synchronization requirements for this to work? Surely reads and writes are not visible immediately on the other side?

No suggestions on how to deal with multiple GPUs, but I'd say it's unlikely that we'll support that any other way than with multiple MPI ranks for quite some time. We could perhaps add an assertion somewhere in the initialization to avoid surprises, but I don't know if that's necessary.

Reply:
> Today I learned! I only knew pinned host memory was required for asynchronous transfers, but not that it could actually be accessed directly on the device. Are there any special synchronization requirements for this to work? Surely reads and writes are not visible immediately on the other side?

Wait, I'm not entirely sure at this point. I remember of something like zero-copy/unified memory, but it may be a specific use case I had when I was working on a Jetson (which is quite "special" in terms of hardware architecture).

Let me check and I will write here.

Reply:
From [CUDA doc](https://docs.nvidia.com/cuda/archive/11.0/cuda-runtime-api/group__CUDART__UNIFIED.html#group__CUDART__UNIFIED) about Unified Virtual Addressing (UVA)

> Automatic Mapping of Host Allocated Host Memory
>
> All host memory allocated through all devices using cudaMallocHost() and cudaHostAlloc() is always directly accessible from all devices that support unified addressing. This is the case regardless of whether or not the flags cudaHostAllocPortable and cudaHostAllocMapped are specified.
>
> The pointer value through which allocated host memory may be accessed in kernels on all devices that support unified addressing is the same as the pointer value through which that memory is accessed on the host. It is not necessary to call cudaHostGetDevicePointer() to get the device pointer for these allocations.

From [Unified Memory or Unified Virtual Addressing?](https://developer.nvidia.com/blog/unified-memory-in-cuda-6/)
> UVA enables “Zero-Copy” memory, which is pinned host memory accessible by device code directly, over PCI-Express, without a memcpy. Zero-Copy provides some of the convenience of Unified Memory, but none of the performance, because it is always accessed with PCI-Express’s low bandwidth and high latency.

We use UVA (its support is currently implicitly required in DLAF), so in the case of this PR zero-copy is used. Since we are accessing memory just one time and not multiple times, it may be a good solution, but we will do some benchmark in the near future to confirm this.

**TL;DR** CUDA Unified Virtual Addressing (UVA) enables pinned host memory to be accessed by the GPU (via zero-copy)

Reply:
> Today I learned! I only knew pinned host memory was required for asynchronous transfers, but not that it could actually be accessed directly on the device. Are there any special synchronization requirements for this to work? Surely reads and writes are not visible immediately on the other side?

I think this is discussed in the Mapped Memory section: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#mapped-memory . The following is stated:

> A block of page-locked host memory can also be mapped into the address space of the device by passing flag cudaHostAllocMapped to `cudaHostAlloc()`

So I am not sure we can assume it's available on the GPU unless the flag is passed.

Reply:
> So I am not sure we can assume it's available on the GPU unless the flag is passed.

From the doc:
>__host__ ​cudaError_t cudaMallocHost ( void** ptr, size_t size )
> Allocates size bytes of host memory that is page-locked and accessible to the device.

This means that the flag is passed.